Fix regression for label scans

pontusmelke opened the pull request:
After switching to the new API there was a regression for scanning nodes by label in the Core API. This PR addresses that PR by:
- Less indirection in the cursor, or rather the cursor knows about the type of the progressor and thus avoiding a megamorphic call site
- More efficient cursor-to-iterator in `GraphDatabaseFacade`

pontusmelke committed:
More efficient iterator for label scans

pontusmelke committed:
Tighter coupling between cursor and progressor

Allow the type of the progressor of the cursor to be known at compile
time. This removes a megmorphic call site and gives big boost in
performance benchmarks.

craigtaverner commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -922,4 +895,69 @@ public GraphPropertiesProxy newGraphPropertiesProxy()
+        {
+            this.cursor = cursor;
+            this.statement = statement;
+            next = -2L;
```
Should be UNINITIALIZED

craigtaverner replied:
Or perhaps the first fetchNext could be called in the constructor to remove the need for this special value?

craigtaverner commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -922,4 +895,69 @@ public GraphPropertiesProxy newGraphPropertiesProxy()
+    private final class NodeCursorResourceIterator implements ResourceIterator<Node>
+    {
+        private static final long UNINITIALIZED = -2L;
+        private static final long NO_ID = -1L;
```
Could use org.neo4j.kernel.impl.store.record.AbstractBaseRecord.NO_ID

craigtaverner commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/NodeLabelIndexCursor.java:
```diff
@@ -109,7 +109,8 @@ public boolean next()
         else
         {
-            return innerNext();
+            LabelScanValueIndexProgressor p = progressor;
```
This change does not seem to do anything? The same code exists in `innerNext`, right?

pontusmelke committed:
Fixes from code review

pontusmelke committed:
Made iterator lazily initialized again

Turns out we need to initialize the iterator lazily because otherwise
it messes up semantics around removals of labels after initialization
of the the iterator has been created.

1. Open iterator for a label scan
2. remove label on a node (now we cannot have already fetched that node in
  step 1
